File: 2022/04/04.rb
# frozen_string_literal: true

path = File.expand_path('../inputs/4.txt', __dir__)

input_ranges = File.readlines(path, chomp: true).map do |elf_assignments|
  elf1, elf2 = elf_assignments.split(',')
  elf1_start, elf1_end = elf1.split('-').map(&:to_i)
  elf2_start, elf2_end = elf2.split('-').map(&:to_i)
  [Range.new(elf1_start, elf1_end), Range.new(elf2_start, elf2_end)]
end

part1 = input_ranges.count do |elf1_range, elf2_range|
  elf1_range.cover?(elf2_range) || elf2_range.cover?(elf1_range)
end

part2 = input_ranges.count do |elf1_range, elf2_range|
  !(elf1_range.last < elf2_range.first || elf2_range.last < elf1_range.first)
end

puts "part 1:  #{part1} "
puts "part 2:  #{part2} "

File: 2022/04/04.py
import re

with open("inputs/4.txt") as file:
    assignments = []
    for elf_assignments in file:
        values = [int(value) for value in re.split(",|-", elf_assignments.strip())]
        assignments.append([values[0], values[1], values[2], values[3]])


def part_1():
    count = 0
    for (elf1_start, elf1_end, elf2_start, elf2_end) in assignments:
        if (elf2_start >= elf1_start and elf2_end <= elf1_end) or (
            elf1_start >= elf2_start and elf1_end <= elf2_end
        ):
            count += 1
    return count


def part_2():
    return sum(
        [
            1
            for (elf1_start, elf1_end, elf2_start, elf2_end) in assignments
            if elf1_end >= elf2_start and elf2_end >= elf1_start
        ]
    )


print(f"Part 1: {part_1()}")
print(f"Part 2: {part_2()}")
